Undo v1.20.2-3 + session checkpoints

mark-undoio opened the pull request:
Extend previous work on the `restart` command and checkpoint facilities to support save/load of sessions in a compatible format to UDB.

This PR introduces the concept of an `undoSession` object, which stores checkpoints and is provided as a context for resolving user-specified times.  This enables it to do appropriate look-up for checkpoints when required and gives us a home for session-management (such as saving and loading the JSON file).

mark-undoio committed:
Move "local checkpoint" management code into an Undo session object.

mark-undoio committed:
Implement save and load of checkpoints with an Undo session file.

This is made compatible with UDB's bookmarks by using the UDB bookmark name as
the "Note" field in Delve. To save Delve checkpoints, the "Note" must be made
unique, so as to be a valid bookmark name in UDB.

gareth-rees commented on pkg/proc/gdbserial/gdbserver.go:
```diff
@@ -167,8 +167,7 @@ type gdbProcess struct {
 
-	localCheckpointLastId int
-	localCheckpoints      map[int]proc.Checkpoint
+	undoSession *undoSession
```
A comment would be useful here explaining the relationship between the GDB process and the Undo session. (For example, does one belong to the other, or use settings loaded from the other?) Also, since this is a pointer it can be `nil`, so you might document the invariant that lets us know when it is non-`nil`

mark-undoio replied:
:+1:

gareth-rees commented on pkg/proc/gdbserial/gdbserver.go:
```diff
@@ -1127,45 +1123,10 @@ func (p *gdbProcess) restartWorker(cctx *proc.ContinueOnceContext, pos string) (
-			// used as an argument to vUDB;goto_time.
-			pos = fmt.Sprintf("%x;%x", bbcount, pc)
+		var err error
+		pos, err = p.undoSession.resolveUserTime(p, pos)
```
How do we know that `p.undoSession` is not `nil` here?

mark-undoio replied:
I'm addressing this, for now, with a top level comment.

gareth-rees commented on pkg/proc/gdbserial/gdbserver.go:
```diff
@@ -1180,14 +1141,7 @@ func (p *gdbProcess) restartWorker(cctx *proc.ContinueOnceContext, pos string) (
-		default:
-			err = p.conn.restart(pos)
-		}
+		err = p.undoSession.travelToTime(p, pos)
```
Similarly here. The invariant seems to be that `p.conn.isUndoServer` if and only if `p.undoSession != nil`. Is that right? If so, could we drop the former and use the latter instead? (Not sure if this is possible.)

mark-undoio replied:
I'm going to investigate unifying this further in future - there's potentially a layering issue but it might be fine.  Still getting my head around how this lines up.

gareth-rees commented on pkg/proc/gdbserial/undo.go:
```diff
@@ -2,19 +2,318 @@ package gdbserial
 	"github.com/undoio/delve/pkg/proc"
 )
 
+// State relating to an Undo "session" - this is runtime state that corresponds with.
```
Something seems to be missing at the end of the sentence.

gareth-rees commented on pkg/proc/gdbserial/undo.go:
```diff
@@ -2,19 +2,318 @@ package gdbserial
 
+// State relating to an Undo "session" - this is runtime state that corresponds with.
+type undoSession struct {
+	checkpointLastId int
+	checkpoints      map[int]proc.Checkpoint
```
Could you add comments documenting these fields?

gareth-rees commented on pkg/proc/gdbserial/undo.go:
```diff
@@ -2,19 +2,318 @@ package gdbserial
+	checkpoints      map[int]proc.Checkpoint
+}
+
+// Create a new undoSession structure, which captures the runtime state relating to an on-disk
+// "session" file - currently only bookmarks are supported.
+func newUndoSession() *undoSession {
```
The comment could be clarified, since nothing in the body of the function relates to an on-disk session file. Or perhaps most of this really belongs in the `undoSession` comment instead.

mark-undoio replied:
:+1:

gareth-rees commented on pkg/proc/gdbserial/undo.go:
```diff
@@ -2,19 +2,318 @@ package gdbserial
+		return errors.New("Checkpoint note cannot be empty.")
+	}
+
+	// Perform matching checking to UDB's bookmark creation code, to ensure Delve notes
+	// translate properly into UDB bookmarks when we save the session.
```
It would be helpful to have a cross-reference to the corresponding code in UDB.

gareth-rees commented on pkg/proc/gdbserial/undo.go:
```diff
@@ -2,19 +2,318 @@ package gdbserial
+// Validate a checkpoint note to ensure easy interopability with UDB bookmarks.
+func validateCheckpointNote(where string) error {
+	if where == "" {
+		return errors.New("Checkpoint note cannot be empty.")
```
I think "must not" would be clearer than "cannot" here and below (compare [RFC 2119](https://www.rfc-editor.org/rfc/rfc2119))

gareth-rees commented on pkg/proc/gdbserial/undo.go:
```diff
@@ -2,19 +2,318 @@ package gdbserial
+		return fmt.Errorf("Checkpoint note cannot start with character: %c", where[0])
+	} else {
+		firstWord := strings.Split(where, " ")[0]
+		reserved := []string{"time", "undo", "redo", "inferior", "start", "annotation",
+			"end", "bookmark", "event", "pc", "wallclock"}
```
I suggest keeping lists like this in alphabetical order (unless there is a reason to use some other order) as this makes it easier to check for missing items or duplicates.

gareth-rees commented on pkg/proc/gdbserial/undo.go:
```diff
@@ -2,19 +2,318 @@ package gdbserial
+	}
+}
+
+// Validate a checkpoint note to ensure easy interopability with UDB bookmarks.
```
Consider documenting the result, i.e. `nil` if the checkpoint note is valid, or an error if not.

gareth-rees commented on pkg/proc/gdbserial/undo.go:
```diff
@@ -2,19 +2,318 @@ package gdbserial
+	if err != nil {
+		return -1, err
+	}
+	cpid := uc.checkpointLastId
+	uc.checkpointLastId++
```
The logic here does not match the field name. I suggest either renaming the field to `checkpointNextId`, or incrementing it before taking the value.

gareth-rees commented on pkg/proc/gdbserial/undo.go:
```diff
@@ -2,19 +2,318 @@ package gdbserial
+
+// Look up a Delve checkpoint structure by time, as represented by the vUDB;get_time serial command.
+func (uc *undoSession) lookupCheckpoint(pos string) (proc.Checkpoint, error) {
+	cpid, _ := strconv.Atoi(pos[1:])
```
Would it make sense to check the first character of `pos` is correct?
What happens here if `pos` is the empty string?

mark-undoio replied:
:+1:

gareth-rees commented on pkg/proc/gdbserial/undo.go:
```diff
@@ -2,19 +2,318 @@ package gdbserial
+	return cpid, nil
+}
+
+// Look up a Delve checkpoint structure by time, as represented by the vUDB;get_time serial command.
```
The comment here does not seem to describe the behaviour of the function.

gareth-rees commented on pkg/proc/gdbserial/undo.go:
```diff
@@ -2,19 +2,318 @@ package gdbserial
+	if err != nil {
+		return "", err
+	}
+	uuids := strings.Split(recording_ids, ";")
```
I suggest checking that this has the expected number of elements here, as you will be looking up `uuids[1]` later.

gareth-rees commented on pkg/proc/gdbserial/undo.go:
```diff
@@ -2,19 +2,318 @@ package gdbserial
+
+	// Translate the loaded Undo bookmarks into Delve checkpoints.
+	for name, position := range s.Bookmarks {
+		cpid := uc.checkpointLastId
+		uc.checkpointLastId++
```
See my note above about the field name vs usage.

gareth-rees commented on pkg/proc/gdbserial/undo.go:
```diff
@@ -2,19 +2,318 @@ package gdbserial
+		s.Bookmarks[name] = time
+	}
+
+	path, err := getSessionPath(p)
+	if err != nil {
+		return err
+	}
```
Might be worth doing this first — if there's no loaded recording then we can skip some work.

mark-undoio replied:
There's always a loaded recording in Delve - no live debugging here - so this can only happen if there's a genuinely unexpected error.

gareth-rees commented on pkg/proc/gdbserial/undo.go:
```diff
@@ -2,19 +2,318 @@ package gdbserial
+		checkpoints = append(checkpoints, cp)
+	}
+	// Sort the checkpoints by descending note length - this is to avoid adding a suffix to a
+	// entries that we've already added with a suffix to. e.g. if we've previously saved this
```
Typo: "added with a suffix to"

gareth-rees commented on pkg/proc/gdbserial/undo.go:
```diff
@@ -2,19 +2,318 @@ package gdbserial
+	}
+	// Sort the checkpoints by descending note length - this is to avoid adding a suffix to a
+	// entries that we've already added with a suffix to. e.g. if we've previously saved this
+	// session with a duplicated checkpoint note called "test" will have extended one to
```
There seems to be something missing between "test" and "will".

gareth-rees commented on pkg/proc/gdbserial/undo.go:
```diff
@@ -156,6 +455,13 @@ func UndoReplay(recording string, path string, quiet bool, debugInfoDirs []strin
 	// to the GDB mapping, not the Linux mapping (the binutils-gdb repo
 	// defines the GDB mapping in include/gdb/signals.def)
 	p.conn.isUndoServer = true
+
+	// Create storage for Undo checkpoints, which (unlike rr) aren't stored in the server.
+	p.undoSession = newUndoSession()
```
See note above — `p.conn.isUndoServer` is redundant with `p.undoSession != nil` so maybe the former could now be dropped. (By dropping redundant fields we avoid the risk of the data structure getting into an inconsistent state.)

mark-undoio replied:
This'll want looking into but I think it's a future thing - the current separation is something we've inherited from the original implementation, it might make sense or not be required at all.

gareth-rees reviewed (approved):
The refactoring of a lot of Undo-specific code into methods on an `undoSession` object seems like a big win. None of the comments are merge-blocking but it would be nice if the discrepancies in a couple of places between comments and code could be resolved.

mark-undoio commented:
@gareth-rees I'm adding some additional commentary.

I'll layer on an additional commit for the checkpoint naming stuff, since naming is part of the original fork (and thus we'll want to squash the fix down into the patchset later).

For the `isUndoServer` vs `undoSession` - I've always been unhappy with this split but haven't yet figured out how to deal with it.  I think it's a bit of a can of worms.

The `gdbConn` level currently needs to know if we're an Undo server but can't access `gdbProcess.undoSession`.  I think can attach *most* of the state down into the `gdbConn` layer instead if we wanted but I'm undecided whether that makes sense in terms of the layering - maybe.

We will still be storing some Undo state in the `gdbProcess` even then, since we cheekily store the recording filename in the `gdbProcess.tracedir` directory.  That might be the least bad option

mark-undoio committed:
Fix naming of checkpointLastId -> checkpointNextId.

This is in line with the actual use of this variable.

mark-undoio committed:
Fix bug in handling zero-argument "restart" command.

Previously we were trying to interpret an empty command as a time, which was not
correct.

mark-undoio commented:
@gareth-rees I think I've addressed all the stuff I'm intending to get into right now - which was most of it.  I'm intending to merge with your existing approval but can revisit anything if there are mistakes / misunderstandings in my implementation of these changes.

I snuck an extra fix (zero-argument handling of the "restart" command) in here but I don't think it needs review as it's really very small.

mark-undoio commented:
This was released as part of `undo-v1.20-4`.
